pkg/reconcilers/deployment_mutator.go

```diff
@@ -0,0 +1,545 @@
+	return func(desiredObj, existingObj client.Object) (bool, error) {
+		existing, ok := existingObj.(*k8sapps.Deployment)
+		if !ok {
+			return false, fmt.Errorf("%T is not a *appsv1.Deployment", existingObj)
+		}
+		desired, ok := desiredObj.(*k8sapps.Deployment)
+		if !ok {
+			return false, fmt.Errorf("%T is not a *appsv1.Deployment", desiredObj)
+		}
+
+		update := false
+
+		// Loop through each option
+		for _, opt := range opts {
+			tmpUpdate := opt(desired, existing)
+			update = update || tmpUpdate
+		}
+
+		return update, nil
+	}
+}
+
+func DeploymentReplicasMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	var existingReplicas int32 = 1
+	if existing.Spec.Replicas != nil {
+		existingReplicas = *existing.Spec.Replicas
+	}
+
+	var desiredReplicas int32 = 1
+	if desired.Spec.Replicas != nil {
+		desiredReplicas = *desired.Spec.Replicas
+	}
+
+	if desiredReplicas != existingReplicas {
+		existing.Spec.Replicas = &desiredReplicas
+		update = true
+	}
+
+	return update
+}
+
+func DeploymentContainerListMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	if len(existing.Spec.Template.Spec.Containers) != len(desired.Spec.Template.Spec.Containers) {
+		existing.Spec.Template.Spec.Containers = desired.Spec.Template.Spec.Containers
+		update = true
+	}
+
+	return update
+}
+
+func DeploymentImageMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	if existing.Spec.Template.Spec.Containers[0].Image != desired.Spec.Template.Spec.Containers[0].Image {
+		existing.Spec.Template.Spec.Containers[0].Image = desired.Spec.Template.Spec.Containers[0].Image
+		update = true
+	}
+
+	return update
+}
+
+func DeploymentVolumesMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	if !reflect.DeepEqual(existing.Spec.Template.Spec.Volumes, desired.Spec.Template.Spec.Volumes) {
+		existing.Spec.Template.Spec.Volumes = desired.Spec.Template.Spec.Volumes
+		update = true
+	}
+
+	return update
+}
+
+func DeploymentVolumeMountsMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	existingContainer := &existing.Spec.Template.Spec.Containers[0]
+	desiredContainer := &desired.Spec.Template.Spec.Containers[0]
+
+	if !reflect.DeepEqual(existingContainer.VolumeMounts, desiredContainer.VolumeMounts) {
+		existingContainer.VolumeMounts = desiredContainer.VolumeMounts
+		update = true
+	}
+
+	return update
+}
+
+func DeploymentServiceAccountMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	if !reflect.DeepEqual(existing.Spec.Template.Spec.ServiceAccountName, desired.Spec.Template.Spec.ServiceAccountName) {
+		existing.Spec.Template.Spec.ServiceAccountName = desired.Spec.Template.Spec.ServiceAccountName
+		update = true
+	}
+
+	return update
+}
+
+func DeploymentLabelsMutator(desired, existing *k8sapps.Deployment) bool {
+	update := false
+
+	if !reflect.DeepEqual(existing.ObjectMeta.Labels, desired.ObjectMeta.Labels) {
```

Comment:
We should use set comparison here and avoid false positives due to ordering.

Reply:
labels are maps, ordering not required. `reflect.DeepEqual` is the appropriate tool to compare maps